### ISSUE TITLE
Banker's rounding on decimalplaces(x,n)

### ISSUE THREAD

**Issue by malthefogsporring** [maths/moodle-qtype_stack#616]
Is there a good reason why significantfigures(x,n) does not perform banker's rounding, but decimalplaces(x,n) does? I.e. 2.05 is rounded to 2.0 instead of 2.1. It seems the main advantage of defining our own STACK functions is we can avoid Maxima's banker's rounding. 

I suggest changing decimalplaces to not perform banker's rounding. Question authors who want to simply truncate numbers can use dispdp(x,n).

[rounding-question.zip](https://github.com/maths/moodle-qtype_stack/files/5084293/rounding-question.zip)

**Comment by shinuito** [comment]
I also dislike the use of bankers rounding in `decimalplaces(x,n)`. Maxima's function `round` doesn't allow one to choose the number of decimal places to round to, and I presume the idea of bankers rounding there was really meant that you round toward the nearest even number 2.5->2 and 1.5->2 etc when the number is really an integer +0.5 . 

Here `decimalplaces(x,n)` is just rounding to the nearest even digit for the number of decimal places required, when the next digit is a 5, which is both peculiar and somewhat tedious when trying to compare answers.

Imagine comparing a students answer to four decimal places but the correct answer to five decimal places can end in a 5 sometimes (assume some randomization in the question). Suppose you'd like to give marks regardless of which way they decide to round up/down, now one would have to check the fifth and fourth decimal place of number to see which way the rounding function behaves, and then add/subtract 1e-4 for a second comparison.

The `NumDecPlaces` comparison in the PRT, I believe does this weird bankers rounding, meaning the correct answers to four decimal places of the decimals 0.12345, 0.12355, 0.12365 are 0.1234, 0.1236, 0.1236. It then makes it seems like this comparison option is less useful and would nearly be better to split it into two comparison options, one that just checks the number of submitted decimal places, and one that does the rounding and comparison, since one might have to manually do these comparisons anyway if one dislikes bankers rounding (which I presumes isn't usual for students to do).

**Comment by lassikor** [comment]
I agree with the suggestion not to use bankers rounding. The different rounding method used in functions decimalplaces() and signigicantfigures() is confusing and I personally don't know a student who would be familiar with the bankers rounding. This makes the use of the anwer test NumDecPlaces quite useless as well, or at least one must be very careful when using it.

**Comment by aharjula** [comment]
Attached is a test script showing how `decimalplaces(x,n)` fails to do banker's rounding, and shows a variant that fails less often. Someone might want to test their questions with the replacement function. I am not entirely sure if one needs to plug it both into the question-variables and into feedback-variables or does it work with just the question-variables currently.

Add this for testing, note that while you can replace internal STACK functions like this through the question-variables it is not something that one should do, as the functions and their use may change during version upgrades and your materials may react badly to that:
```
/* Multi-line expressiosn only work since 4.3, if older write to a single line. */
decimalplacesfun(ex, n, dispdps) := block([ex2],
    ex2:ev(float(round(10^n*rat(float(ex)))/(10^n)), lg=logbasesimp, simp),
    if dispdps then ex2:displaydp(ex2, n),
    return(ex2)
    );
```
I would expect that to fix Lassis example picture, and someone might even figure out how to switch to more rational rounding logic by patching that although I would not recommend doing that. Also, note that this replacement function relies heavily on magic-numbers like `ratepsilon` so it might break down unexpectedly unless we fix those numbers. Note that this has not been tested with all forms of inputs and is expected to be more expensive to evaluate. Needs testing to check if too expensive and if it works where needed.

Note that in the example file it also has `ratprint=false` as an evaluation-flag but for whatever reasons that is currently forbidden, luckily it is not needed for testing this, for real use would be nice to include it.

[rounding.mac.txt](https://github.com/maths/moodle-qtype_stack/files/5517030/rounding.mac.txt)

As to the bigger issue of rounding, I would prefer that STACK would have a single setting (needs to be a variable that we can change in the middle of the logic) for selecting the rounding to use and we would not need to know which rounding specific functions use, as every function would use the same defined one. Naturally, it would then mean that we would need to define the rounding logic and types we support. While it might seem odd that the default in Maxima is what it is there is actually a reason for it, banker's rounding is unbiased, which is often a very nice feature and IEEE float arithmetic specs actually prefer it. But naturally, we can change the logic as much as we want.

In any case, if we build other rounding logics into the system we really need to test them for funny float behaviour and if need be add expensive rational number logic to try to fix things although we can not fix all cases like that. Luckily, rounding is often just the last step and therefore does not happen often so we can afford to spend some more time doing it.
